_posts/2016/2016-12-20-cqrs-and-elasticsearch.md

```diff
@@ -0,0 +1,228 @@
+---
+layout: post
+title: "CQRS and Elasticsearch"
+date: "2016-12-20 09:00:00"
+image: blog-header/searching.jpg
+categories: tech-stories
+authors: ["Jens"]
+---
+
+After having explained how we use [Multitenancy and Elasticsearch][prev-blog-post], we will continue our series of search-related blog posts by looking at a software architecture pattern called `Command and Query Responsibility Segregation` ([CQRS][fowler-cqrs]) and how we embrace it for managing and accessing product data in our ePages shops.
+
+## Different Product Data Models
+
+In our ePages software we identified two fundamentally different usage scenarios when dealing with product data, each coming with its own data model.
+
+### The Write Model
+
+A merchant uses the back office application to manage all aspects of her shop, in particular it's product data.
+This product data is stored in a relational database, where transactions and database constraints guarantee consistency and integrity.
+A simplified relational data model looks like this:
+
+{% image blog/blog-cqrs-elasticsearch-database.png %}
+
+### The Read Model
+
+The storefront application renders product data for the customers visiting a merchant's shop.
+It is displaying a list of products as a result of a customer browsing a category or using the search functionality of the shop.
+When navigating to a product detail page, more product-related data is presented to the customer.
+
+Especially the need for full-text searching and faceted filtering (also called *after-search navigation*), influenced the choice to use search technology, namely [Elasticsearch][es], as the only source for accessing product data from the storefront.
+A simplified JSON document of a particular product to be indexed by Elasticsearch looks like this:
+
+{% highlight javascript %}
+{
+    "sku" : "334163-2",
+    "name" : "Sneakers",
+    "description" : "best running experience",
+    "price" : 49.99,
+    "availability" : "IN_STOCK",
+    "images" : [{
+        "name" : "front",
+        "uri" : "https://myshop.com/images/334163-2/front.jpg"
+    }, {
+        "name" : "back",
+        "uri" : "https://myshop.com/images/334163-2/back.jpg"
+    }]
+}
+{% endhighlight %}
+
+The *availability* state is a computed value based on the *stock* and *threshold* numbers.
+
+This full set of attributes is only used when rendering the product detail page.
+Lists of products only show a subset of these attributes, hiding *sku*, *description* and all but one *image*.
+
+## Command and Query Responsibility Segregation
```

Comment:
It's a known term in software architecture, hence the abbreviation CQRS. Thus I would keep it like that.